sdk/go/property/path.go

```diff
@@ -0,0 +1,215 @@
+// Copyright 2016-2024, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package property
+
+import (
+	"fmt"
+)
+
+// Path provides access and alteration methods on [Value]s.
+//
+// Paths are composed of [PathSegment]s, which can be one of:
+//
+// - [KeySegment]: For indexing into [Map]s.
+// - [IndexSegment]: For indexing into [Array]s.
+type Path []PathSegment
+
+func (p Path) Get(v Value) (Value, PathApplyFailure) {
+	for _, segment := range p {
+		var err PathApplyFailure
+		v, err = segment.apply(v)
+		if err != nil {
+			return Value{}, err
+		}
+	}
+	return v, nil
+}
+
+// Set a value.
+//
+// Set does not src in place, rather producing a copy.
```

Comment:
```suggestion
// Set does not change src in place, rather producing a copy.
```

I'm also curious about this decision.  I get that immutable data structures are sometimes convenient, and easier to reason about, but it is more expensive, especially for large arrays/maps.

Reply:
It's also not very Go-ish. I like immutable, but I do question if it will actually make things easier in Go?